Session 0 Ruby Questions

ronithrnair opened the pull request:
Session 0 Ruby questions completed (I think)

abhishekkumar2718 commented:
As you can see by the failing check for Session 0, ruby is _not_ done - try running the tests locally to check if your solutions are correct

ronithrnair commented:
Sorry, I forgot to merge a branch into the session_0 branch

ronithrnair commented:
> As you can see by the failing check for Session 0, ruby is _not_ done - try running the tests locally to check if your solutions are correct

The tests ran properly now that I've corrected my code.

abhishekkumar2718 commented on session_0/fizz_buzz.rb:
```diff
@@ -36,6 +36,27 @@
+     raise ArgumentError # TODO
+    end
+   a = []
+   for i in 1..n do
```
Do you think map function can be of use here?

https://www.rubyguides.com/2018/10/ruby-map-method/

abhishekkumar2718 commented on session_0/item.rb:
```diff
@@ -23,14 +23,28 @@ def initialize(name: '', price: 0, quantity: 1, category_id: 4,
   # discount deadline has been crossed or not.
   def discounted?
-    raise NotImplementedError # TODO
+    if (Time.now > @discount_deadline) 
```
- Use `discount_deadline` as we have define attr accessor for it.

```ruby
if condition
  return false
else
  return true
end
```

is same as:
```ruby
return !condition
```

abhishekkumar2718 commented on session_0/item.rb:
```diff
@@ -23,14 +23,28 @@ def initialize(name: '', price: 0, quantity: 1, category_id: 4,
   # TODO: Implement instance method 'current_price'
 
+  def current_price
+    if discounted? then
```
Don't use the `then` keyword, it does not add to the readability.

abhishekkumar2718 commented on session_0/item.rb:
```diff
@@ -41,5 +55,16 @@ def discounted?
   # Note: If there are no items for category, stock price for category
   # should be zero.
   def self.stock_price_by_category(items)
+    total_category_price = {
```
What if there are 100 categories? Would writing a hash by hand still make sense?

abhishekkumar2718 commented on session_0/word_processing.rb:
```diff
@@ -2,13 +2,16 @@
 # and does not return any value.
 def lower_case!(words)
-  raise NotImplementedError # TODO
+  for a in words
```
Use `words.each` instead of a `for` loop

abhishekkumar2718 commented on session_0/word_processing.rb:
```diff
@@ -34,11 +37,32 @@ def words_with_prefix(prefix, words)
-  raise NotImplementedError # TODO
+  size = [word_1.size,word_2.size].min # TODO
+  score = 0
+  for i in 0..size-1
```
What are you iterating over in the for loop?

ronithrnair replied:
I was trying to extract each individual character in the word in the same way one does in C++. Is it possible to use word_1.each  to get individual characters?

abhishekkumar2718 replied:
Experiment and find out

ronithrnair commented:
I changed Fizz Buzz to use map and I removed all the 'then's. I committed these changes - What do I do now ?  As in , can you see the changes that I have made?  Also I don't know how to resolve the other issue -  the writing by hash one.

ronithrnair commented on session_0/fizz_buzz.rb:
```diff
@@ -36,6 +36,23 @@
+     if str== "" 
+     str << i.to_s
+     end
+     k = str
```
It was a mistake, really. I forgot to remove the statement from a previous version of the code. I now realise that the presence of an assignment statement somehow made the program work. From what I understand it has to do with the fact that assignment statements return the value of whatever value it's assigning.

ronithrnair commented:
I think I made all the changes.

abhishekkumar2718 commented on session_0/word_processing.rb:
```diff
@@ -34,11 +35,45 @@ def words_with_prefix(prefix, words)
-  raise NotImplementedError # TODO
+  size = [word_1.size,word_2.size].min # TODO
+  score = 0
+  for i in 0..size-1
```
See if you can use `find` function here: https://apidock.com/ruby/Enumerable/find
